haskell-src/Concordium/Types/Parameters.hs

```diff
@@ -1,23 +1,393 @@
+    -- that may be created in one block.
+    cpAccountCreationLimit,
+    -- |Reward parameters.
+    cpRewardParameters,
+    -- |Foundation account index.
+    cpFoundationAccount,
+    -- |Parameters for baker pools. Prior to P4, this is just the minimum stake threshold
+    -- for becoming a baker.
+    cpPoolParameters,
+    EChainParameters (..),
+    ChainParameters,
+    putChainParameters,
+    getChainParameters,
+
+    -- * Finalization parameters
+    FinalizationParameters (..),
+    putFinalizationParametersGD3,
+    getFinalizationParametersGD3,
+
+    -- * Delegation helpers
+    DelegationChainParameters (..),
+    delegationChainParameters,
+
+    -- * Authorizations version
+
+    -- |Version of the authorizations structure.
+    --
+    -- * 'AuthorizationsVersion0' ('ChainParametersV0').
+    -- * 'AuthorizationsVersion1' ('ChainParametersV1', 'ChainParametersV2'): add access structures
+    --   for cooldown parameters and time parameters
+    AuthorizationsVersion (..),
+    -- |Singleton type associated with 'AuthorizationsVersion'.
+    SAuthorizationsVersion (..),
+    -- |The authorizations version associated with a chain parameters version.
+    authorizationsVersionFor,
+    -- |The authorizations version associated with a chain parameters version (types).
+    AuthorizationsVersionFor,
+    -- |The authorizations version associated with a chain parameters version (singletons).
+    sAuthorizationsVersionFor,
+    -- |The authorizations version associated with a protocol version.
+    authorizationsVersionForPV,
+    -- |The authorizations version associated with a protocol version (types).
+    AuthorizationsVersionForPV,
+    -- |The authorizations version associated with a protocol version (singletons).
+    sAuthorizationsVersionForPV,
+    IsAuthorizationsVersion,
+    withIsAuthorizationsVersionFor,
+    withIsAuthorizationsVersionForPV,
+    -- |Whether cooldown parameters are updatable for an 'AuthorizationsVersion'.
+    supportsCooldownParametersAccessStructure,
+    -- |Whether cooldown parameters are updatable for an 'AuthorizationsVersion' (types).
+    SupportsCooldownParametersAccessStructure,
+    -- |Whether cooldown parameters are updatable for an 'AuthorizationsVersion' (singletons).
+    sSupportsCooldownParametersAccessStructure,
+    -- |Whether time parameters are supported for an 'AuthorizationsVersion'.
+    supportsTimeParameters,
+    -- |Whether time parameters are supported for an 'AuthorizationsVersion' (types).
+    SupportsTimeParameters,
+    -- |Whether time parameters are supported for an 'AuthorizationsVersion' (singletons).
+    sSupportsTimeParameters,
+
+    -- * Defunctionalisation symbols
+    PTElectionDifficultySym0,
+    PTTimeParametersSym0,
+    PTMintPerSlotSym0,
+    PTTimeoutParametersSym0,
+    PTMinBlockTimeSym0,
+    PTBlockEnergyLimitSym0,
+    PTCooldownParametersAccessStructureSym0,
+    PTFinalizationProofSym0,
```

Comment:
Explain in detail and on examples what defunctionalization symbols are. What are they for, what problems do they solve.

Reply:
I have put documentation before the splice. Please check that it is adequate.